File: Examples/BinaryExampleAPK/BadPracticesAppAndroid/app/src/main/cpp/unsafe_capi.c
/*
 * unsafe_capi.c
 * Equivalente Android (NDK) de ObjC/UnsafeCAPI.m — UnsafeCAPIRule.
 * APIs C inseguras compiladas dentro del .so (sobreviven al binario).
 */
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

void cacomi_unsafe_capi_demo(const char *userInput) {
    char buf[16];
    // BAD: strcpy sin verificar limites // CACOMI-EXPECT: UnsafeCAPIRule
    strcpy(buf, userInput);

    char fmt[32];
    // BAD: sprintf sin limites (usar snprintf) // CACOMI-EXPECT: UnsafeCAPIRule
    sprintf(fmt, "user=%s", userInput);

    char line[64];
    // BAD: gets es inherentemente inseguro // CACOMI-EXPECT: UnsafeCAPIRule
    gets(line);

    // BAD: system() con cadena (inyeccion de comandos) // CACOMI-EXPECT: UnsafeCAPIRule
    system("ls /tmp");

    // BAD: rand() es un PRNG debil // CACOMI-EXPECT: UnsafeCAPIRule
    int r = rand();

    (void)buf; (void)fmt; (void)line; (void)r;
}

File: Examples/BinaryExampleIPA/SourceCodeForTesting/BadPracticesIPA/BadPracticesIPA/CacomiTestApp/ObjC/UnsafeCAPI.m
//
//  UnsafeCAPI.m
//  CacomiTestApp
//
//  Objective-C fixtures for UnsafeCAPIRule.
//

#import <Foundation/Foundation.h>
#import <string.h>
#import <stdio.h>
#import <stdlib.h>

void cacomi_unsafe_capi_demo(const char *userInput) {
    char buf[16];
    // CACOMI-EXPECT[UnsafeCAPIRule|critical]: strcpy without bounds check
    strcpy(buf, userInput);

    char fmt[32];
    // CACOMI-EXPECT[UnsafeCAPIRule|critical]: sprintf without bounds (use snprintf)
    sprintf(fmt, "user=%s", userInput);

    char line[64];
    // CACOMI-EXPECT[UnsafeCAPIRule|critical]: gets is inherently unsafe
    gets(line);

    // CACOMI-EXPECT[UnsafeCAPIRule|critical]: literal "system(" sink, function unavailable on iOS
    const char *sysSink = "system(\"ls /tmp\")";
    (void)sysSink;

    // CACOMI-EXPECT[UnsafeCAPIRule|high]: rand() is a weak PRNG
    int r = rand();
    (void)r;
    (void)buf;
    (void)fmt;
    (void)line;
}
